pact/Pact/Core/IR/Desugar.hs

```diff
@@ -1453,33 +1461,34 @@ checkImplements i defs moduleName ifaceName = do
+    IfDfun (IfDefun (Arg ifdName ifRty _) ifArgs _) ->
+      case find (\df -> ifdName == defName df) defs of
+        Just (Dfun (Defun (Arg funName funRty _) funArgs _ _)) ->
+          when (weakArgsNeq funArgs ifArgs || funRty /= ifRty) $
+            throwDesugarError (ImplementationError moduleName ifaceName funName) i
         Just d ->
           throwDesugarError (ImplementationError moduleName ifaceName  (defName d)) i
         Nothing ->
-          throwDesugarError (NotImplemented moduleName ifaceName (_ifdName ifd)) i
-    IfDCap ifd ->
-      case find (\df -> _ifdcName ifd == defName df) defs of
-        Just (DCap v) -> do
-          unless (checkMetaMatches (_dcapMeta v) (_ifdcMeta ifd)) $
-             throwDesugarError (ImplementationError moduleName ifaceName $ ": defcap mismatch for " <> _dcapName v) i
-          when (_dcapArgs v /= _ifdcArgs ifd || _dcapRType v /= _ifdcRType ifd) $
-            throwDesugarError (ImplementationError moduleName ifaceName (_dcapName v)) i
+          throwDesugarError (NotImplemented moduleName ifaceName ifdName) i
+    IfDCap (IfDefCap (Arg ifdcName ifdcRty _) ifdcArgs' ifdcMeta' _) ->
+      case find (\df -> ifdcName == defName df) defs of
+        Just (DCap (DefCap (Arg dcapName dcapRty _) dcapArgs' _ dcapMeta' _)) -> do
+          unless (checkMetaMatches dcapMeta' ifdcMeta') $
+             throwDesugarError (ImplementationError moduleName ifaceName $ ": defcap mismatch for " <> dcapName) i
+          when (weakArgsNeq dcapArgs' ifdcArgs' || dcapRty /= ifdcRty) $
+            throwDesugarError (ImplementationError moduleName ifaceName dcapName) i
         Just d -> throwDesugarError (ImplementationError moduleName ifaceName  (defName d)) i
         Nothing ->
-          throwDesugarError (NotImplemented moduleName ifaceName (_ifdcName ifd)) i
-    IfDPact ifd ->
-      case find (\df -> _ifdpName ifd == defName df) defs of
-        Just (DPact v) ->
-          when (_dpArgs v /= _ifdpArgs ifd || _dpRetType v /= _ifdpRType ifd) $
-          throwDesugarError (ImplementationError moduleName ifaceName (_ifdpName ifd)) i
-        Just _ ->  throwDesugarError (ImplementationError moduleName ifaceName (_ifdpName ifd)) i
-        Nothing -> throwDesugarError (NotImplemented moduleName ifaceName (_ifdpName ifd)) i
-
+          throwDesugarError (NotImplemented moduleName ifaceName ifdcName) i
+    IfDPact (IfDefPact (Arg ifdpName ifdpRty _) ifdpArgs' _) ->
+      case find (\df -> ifdpName == defName df) defs of
+        Just (DPact (DefPact (Arg _ dpRty _) dpArgs' _ _)) ->
+          when (weakArgsNeq dpArgs' ifdpArgs' || dpRty /= ifdpRty) $
+          throwDesugarError (ImplementationError moduleName ifaceName ifdpName) i
+        Just _ ->  throwDesugarError (ImplementationError moduleName ifaceName ifdpName) i
+        Nothing -> throwDesugarError (NotImplemented moduleName ifaceName ifdpName) i
+
+  weakArgsNeq a b = fmap void a /= fmap void b
```

Comment:
doc comment on this function would be helpful. It takes a couple seconds to work out what `weakArgsNeq` is doing here and convince myself it's right. Maybe rename to `argsHaveSameName`?